drivers/net/mvpp2/mrvl_ptp.c

```diff
@@ -0,0 +1,431 @@
+{
+	struct timespec ts;
+	int delta;
+
+	/* The packet timestamp consists of 2 bits of seconds and 30 bits of
+	 * nanoseconds.  We use our stored timestamp (tai->ts) to form a full
+	 * timestamp, and we must read the seconds exactly once.
+	 */
+	if (clock_gettime(tai, &ts)) {
+		return -1;
+	}
+
+	/* Calculate the delta in seconds between our stored timestamp and
+	 * the value read from the queue. Allow timestamps one second in the
+	 * past, otherwise consider them to be in the future.
+	 */
+	delta = ((pkt_ts >> 30) - (ts.tv_sec & 3)) & 3;
+	if (delta == 3)
+		delta -= 4;
+	ts.tv_sec += delta;
+
+	*full_ts = (uint64_t)ts.tv_sec * 1000000000 + ts.tv_nsec;
+
+	return 0;
+}
+
+static void mvpp2_phc_get_current_time(void *arg)
+{
+	struct mrvl_priv *priv = arg;
+	struct timespec ts;
+	int rc;
+
+	if (priv->tai == CLOCK_INVALID)
+		return;
+
+	clock_gettime(priv->tai, &ts);
+
+	rc = rte_eal_alarm_set(2*US_PER_S, mvpp2_phc_get_current_time,
+			(void *)priv);
+	if (rc != 0) {
+		MRVL_LOG(ERR, "Failed to re-schedule PTP alarm\n");
+		priv->scheduled_get_time_alarm = false;
+	}
+}
+
+int mvpp2_schedule_phc_alarm(struct mrvl_priv *priv)
+{
+	if (priv->scheduled_get_time_alarm)
+		return 0;
+
+	int ret = rte_eal_alarm_set(2*US_PER_S, mvpp2_phc_get_current_time,
+			(void *)priv);
+	if (ret == 0) {
+		priv->scheduled_get_time_alarm;
+	}
+	return ret;
+}
+
+void mvpp2_cancel_phc_alarm(struct mrvl_priv *priv)
+{
+	if (priv->scheduled_get_time_alarm) {
+		rte_eal_alarm_cancel(mvpp2_phc_get_current_time, (void *)priv);
+		priv->scheduled_get_time_alarm = false;
+	}
+}
+
+/* PTPAction */
+enum mvpp22_ptp_action {
+	MVPP22_PTP_ACTION_NONE = 0,
+	MVPP22_PTP_ACTION_FORWARD = 1,
+	MVPP22_PTP_ACTION_CAPTURE = 3,
+	/* The following have not been verified */
+	MVPP22_PTP_ACTION_ADDTIME = 4,
+	MVPP22_PTP_ACTION_ADDCORRECTEDTIME = 5,
+	MVPP22_PTP_ACTION_CAPTUREADDTIME = 6,
+	MVPP22_PTP_ACTION_CAPTUREADDCORRECTEDTIME = 7,
+	MVPP22_PTP_ACTION_ADDINGRESSTIME = 8,
+	MVPP22_PTP_ACTION_CAPTUREADDINGRESSTIME = 9,
+	MVPP22_PTP_ACTION_CAPTUREINGRESSTIME = 10,
+};
+
+/* PTPPacketFormat */
+enum mvpp22_ptp_packet_format {
+	MVPP22_PTP_PKT_FMT_PTPV2 = 0,
+	MVPP22_PTP_PKT_FMT_PTPV1 = 1,
+	MVPP22_PTP_PKT_FMT_Y1731 = 2,
+	MVPP22_PTP_PKT_FMT_NTPTS = 3,
+	MVPP22_PTP_PKT_FMT_NTPRX = 4,
+	MVPP22_PTP_PKT_FMT_NTPTX = 5,
+	MVPP22_PTP_PKT_FMT_TWAMP = 6,
+};
+
+#define MVPP22_PTP_ACTION(x)		(((x) & 15) << 0)
+#define MVPP22_PTP_PACKETFORMAT(x)	(((x) & 7) << 4)
+#define MVPP22_PTP_MACTIMESTAMPINGEN	BIT(11)
+#define MVPP22_PTP_TIMESTAMPENTRYID(x)	(((x) & 31) << 12)
+#define MVPP22_PTP_TIMESTAMPQUEUESELECT	BIT(18)
+
+#define MVPP22_PTP_DESC_MASK_LOW  0xfff
+#define PTP_CLASS_V1    0x01 /* protocol version 1 */
+#define PTP_CLASS_V2    0x02 /* protocol version 2 */
+#define PTP_CLASS_VMASK 0x0f /* max protocol version is 15 */
+
+static inline void mvpp2_txdesc_clear_ptp(u32 *ptp_desc)
+{
+	*ptp_desc &= rte_cpu_to_le_32(~MVPP22_PTP_DESC_MASK_LOW);
+}
+
+inline bool mvpp2_tx_hw_tstamp(
```

Comment:
here too: either move it to header or make it non-inline.

Reply:
non-inline, but I'd better move ol_flag check outside

Reply:
maybe it should be inline. these functions are being called frequently.